File: src/main/java/de/byjoker/myjfql/command/ListCommand.java
package de.byjoker.myjfql.command;

import de.byjoker.myjfql.core.MyJFQL;
import de.byjoker.myjfql.database.Database;
import de.byjoker.myjfql.database.DatabaseAction;
import de.byjoker.myjfql.database.DatabaseService;
import de.byjoker.myjfql.database.Table;
import de.byjoker.myjfql.util.Sorter;

import java.util.*;
import java.util.stream.Collectors;

@CommandHandler
public class ListCommand extends Command {

    public ListCommand() {
        super("list", Arrays.asList("COMMAND", "TABLES", "DATABASES", "ORDER", "FROM", "LIMIT"));
    }

    @Override
    public void handleCommand(CommandSender sender, Map<String, List<String>> args) {
        final DatabaseService databaseService = MyJFQL.getInstance().getDatabaseService();

        if (sender.getSession() == null) {
            sender.sendError("Session of this user is invalid!");
            return;
        }

        if (args.containsKey("DATABASES")) {
            List<String> databases = databaseService.getDatabases().stream().filter(database -> sender.allowed(database.getId(), DatabaseAction.READ))
                    .map(Database::getName)
                    .collect(Collectors.toList());

            if (args.containsKey("LIMIT")) {
                int limit;

                try {
                    limit = formatInteger(args.get("LIMIT"));
                } catch (Exception ex) {
                    sender.sendError("Unknown or undefined limit!");
                    return;
                }

                if (limit <= 0) {
                    sender.sendError("Limit is too small!");
                    return;
                }

                if (databases.size() > limit) {
                    databases = databases.stream().limit(limit).collect(Collectors.toList());
                }
            }

            if (args.containsKey("ORDER")) {
                Sorter.Order order;

                try {
                    order = Sorter.Order.valueOf(Objects.requireNonNull(formatString(args.get("ORDER"))).toUpperCase());
                } catch (Exception ex) {
                    sender.sendError("Unknown or undefined sort order!");
                    return;
                }

                sender.sendResult(Sorter.sortList(databases, order), new String[]{"Database"});
                return;
            }


            sender.sendResult(databases, new String[]{"Database"});
            return;
        }

        if (args.containsKey("TABLES")) {
            List<String> tables = new ArrayList<>();

            if (!args.containsKey("FROM")) {
                List<String> finalTables = tables;
                databaseService.getDatabases().stream().filter(database -> sender.allowed(database.getId(), DatabaseAction.READ)).forEach(database -> {
                    finalTables.addAll(database.getTables().stream().map(Table::getName).collect(Collectors.toList()));
                });
            } else {
                final String identifier = formatString(args.get("FROM"));

                if (!databaseService.existsDatabaseByIdentifier(identifier)) {
                    sender.sendError("Database was not found!");
                    return;
                }

                final Database database = databaseService.getDatabaseByIdentifier(identifier);

                if (!sender.allowed(database.getId(), DatabaseAction.READ)) {
                    sender.sendForbidden();
                    return;
                }

                tables.addAll(database.getTables().stream().map(Table::getName).collect(Collectors.toList()));
            }

            if (args.containsKey("LIMIT")) {
                int limit;

                try {
                    limit = formatInteger(args.get("LIMIT"));
                } catch (Exception ex) {
                    sender.sendError("Unknown or undefined limit!");
                    return;
                }

                if (limit <= 0) {
                    sender.sendError("Limit is too small!");
                    return;
                }

                if (tables.size() > limit) {
                    tables = tables.stream().limit(limit).collect(Collectors.toList());
                }
            }

            if (args.containsKey("ORDER")) {
                Sorter.Order order;

                try {
                    order = Sorter.Order.valueOf(Objects.requireNonNull(formatString(args.get("ORDER"))).toUpperCase());
                } catch (Exception ex) {
                    sender.sendError("Unknown or undefined sort order!");
                    return;
                }

                sender.sendResult(Sorter.sortList(tables, order), new String[]{"Table"});
                return;
            }

            sender.sendResult(tables, new String[]{"Table"});
            return;
        }

        sender.sendSyntax();
    }

}

File: src/main/java/de/byjoker/myjfql/command/ListCommand.kt
package de.byjoker.myjfql.command

import de.byjoker.myjfql.core.MyJFQL
import de.byjoker.myjfql.database.DatabasePermissionLevel
import de.byjoker.myjfql.database.RelationalTableEntry
import de.byjoker.myjfql.util.ResultType
import org.jline.reader.ParsedLine

@CommandHandler
class ListCommand : Command("list", listOf("COMMAND", "TABLES", "DATABASES")) {

    override fun execute(sender: CommandSender, args: Map<String, List<String>>) {
        val databaseService = MyJFQL.getInstance().databaseService

        if (args.containsKey("DATABASES")) {
            sender.sendResult(databaseService.databases.filter { database ->
                sender.allowed(
                    database.id,
                    DatabasePermissionLevel.READ
                )
            }.map { database ->
                RelationalTableEntry().append("name", database.name).append("type", database.type)
            }, mutableListOf("name", "type"), ResultType.RELATIONAL)
            return
        }

        if (args.containsKey("TABLES")) {
            val database = sender.session.getDatabase(databaseService)

            if (database == null) {
                sender.sendError("No database is in use for this user!")
                return
            }

            if (!sender.allowed(database.id, DatabasePermissionLevel.READ_WRITE)) {
                sender.sendForbidden()
                return
            }

            sender.sendResult(database.tables.map { table ->
                RelationalTableEntry().append("name", table.name).append("type", table.type)
            }, listOf("name", "type"), ResultType.RELATIONAL)
            return
        }

        sender.sendSyntax()
    }

    override fun complete(sender: CommandSender, line: ParsedLine): MutableList<String>? {
        sender.session ?: return null

        val args = line.line().uppercase()

        return when {
            !args.contains(" TABLES") && !args.contains(" DATABASES") -> {
                mutableListOf("tables", "databases")
            }
            else -> {
                null
            }
        }
    }
}
